HttpServer connection lifecycle rework - fixes #2508

vietj opened the pull request:
Currently the `Http1xServerConnection` class handles all HTTP messages in a single queue. This refactoring removes the global queue and instead uses a buffer queue in each `HttpServerRequest` implementation. Pipelined requests are now handled in the connection as a linked queue of `HttpServerRequest` where the head is the request sending a response and the tail is the request receiving the HTTP content buffers. In normal case the _head_ and _tail_ are the same request, in pipelined case a request may have been fully received but the response not yet sent and thus _head_ → _tail_.

This refactor is required by https://github.com/vert-x3/wiki/wiki/Vert.x-Roadmap#stream-improvements .

tsegismont commented on src/main/java/io/vertx/core/http/impl/Http1xServerConnection.java:
```diff
@@ -21,14 +21,7 @@
-import io.netty.handler.codec.http.HttpUtil;
-import io.netty.handler.codec.http.HttpVersion;
-import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.codec.http.*;
```
Can we update the [code style guidelines](https://github.com/vert-x3/wiki/wiki/Vert.x-code-style-guidelines) to settle on on thresholds for wildcards usage in imports and static imports?

I'm not saying there's an issue with **your** setup. Just annoying to have a lot of changes like this in pull requests.

vietj replied:
let's not be that much regarding :-) , it's my default setup - I agree that those should be left out when submitting a PR and/or handled separately

tsegismont commented:
Tried the branch on my laptop, `Http1xTest.testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived` fails

tsegismont commented:
Is this the PR which you thought would conflict with #2505 ?

Apparently there's just one file conflicting. Does not look hard to resolve.

Or do you have other PRs not sent yet?

vietj commented:
I don't have any other PR, that's why I said in your previous PR (that you cancelled), that I would have overwritten the changes in this PR and removed the deprecated usages

vietj commented:
can you give me the details of the `Http1xTest.testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived` failure

vietj commented:
I think the test `Http1xTest.testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived` is not related to this PR because it mainly test `HttpClient` and uses a `NetServer` not an `HttpServer`. Either the test is faulty (it can have a race) or there is an issue in the client.

tsegismont commented:
@vietj this is what I got:

```
testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived(io.vertx.test.core.Http1xTest)  Time elapsed: 120.022 sec  <<< ERROR!
java.lang.IllegalStateException: Timed out in waiting for test complete
	at io.vertx.test.core.AsyncTestBase.await(AsyncTestBase.java:127)
	at io.vertx.test.core.AsyncTestBase.await(AsyncTestBase.java:115)
	at io.vertx.test.core.Http1xTest.testCloseTheConnectionAfterResetBeforeResponseReceived(Http1xTest.java:3570)
	at io.vertx.test.core.Http1xTest.testCloseTheConnectionAfterResetBeforeKeepAliveResponseReceived(Http1xTest.java:3478)
```

vietj commented:
@tsegismont I just mean that this test can also fail in master, so it's not related to the PR (I've already fixed races in this test recently and I think then there are still issues)

vietj commented:
@tsegismont I've added an extra commit - when we did a review today I told you I found dubious this:

```java
  synchronized void responseComplete() {
    ...
    HttpServerRequestImpl request = pendingResponse;
    pendingResponse = null;
    HttpServerRequestImpl next = request.nextRequest();
    if (request.isEnded() && next != null) {   // <- this line
      // Handle pipelined request
      handleNext(next);
    }
  }
```

I wrote a test that makes it fail (`testBeginPipelinedRequestByResponseSentBeforeRequestCompletion`) and changed to:

```java
  synchronized void responseComplete() {
    ...
    HttpServerRequestImpl request = pendingResponse;
    pendingResponse = null;
    HttpServerRequestImpl next = request.nextRequest();
    if (next != null) {   // <- this line
      // Handle pipelined request
      handleNext(next);
    }
  }
```

tsegismont commented on src/main/java/io/vertx/core/http/impl/HttpHandlers.java:
```diff
@@ -41,6 +41,25 @@ public HttpHandlers(
     this.exceptionHandler = exceptionHandler;
   }
 
+  Handler<HttpServerRequest> requestHandler() {
```
Having this logic here is confusing. For HTTP/1 this method is invoked when the server connection is created. For HTTP/2, the raw `requestHandler` is read.

`HttpHandlers` should remain a data class and this logic be moved to the `Http1xServerConnection` constructor.

vietj replied:
sounds a better thing to do indeed.

tsegismont commented:
@vietj other than the minor concerns, it would be great if we had some Javadoc explaining (just a summary) how this works.

vietj commented on src/main/java/io/vertx/core/http/impl/Http1xServerConnection.java:
```diff
@@ -76,127 +69,177 @@
-  private boolean paused;
-  private boolean sentCheck;
-  private boolean queueing;
+  private HttpServerRequestImpl pendingRequest;  // tail
```
this is actually not the last received, it is clear as soon as the request is ended, that's why it's named _pending_, perhaps it should be named rather something like _requestInProgress_

tsegismont replied:
Well, when it is not `null`, it is the last received :)

But right, `requestInProgress` sounds good to me.

vietj commented on src/main/java/io/vertx/core/http/impl/Http1xServerConnection.java:
```diff
@@ -76,127 +69,177 @@
-  private boolean sentCheck;
-  private boolean queueing;
+  private HttpServerRequestImpl pendingRequest;  // tail
+  private HttpServerRequestImpl pendingResponse; // head
```
it would be incorrect because a request can still be in process and a response be sent already, think an handler that receives a long body but sent a response before having received fully the request. Like above, this is actually the response being sent, so like above I would suggest something like _responseInProgress_

vietj replied:
I will add more general doc about the behavior of state machine
